### ISSUE TITLE
Cannot call SQL Server stored procedure on HSQLDB

### ISSUE THREAD

**Issue by jesset-twosigma** [jOOQ/jOOQ#7673]
### Expected behavior and actual behavior:
Expected:
Call HSQLDB stored procedure and get the result.

Actual:
```
org.jooq.exception.DataAccessException: SQL [select "dbo"."get_unique_ids"(cast(? as varchar(32672)), cast(? as int), cast(? as bigint)) from (select 1 as dual from information_schema.system_users limit 1) as dual]; user lacks privilege or object not found: ts_allocate_unique_long_id_batch_sp in statement [select "dbo"."get_unique_ids"(cast(? as varchar(32672)), cast(? as int), cast(? as bigint)) from (select 1 as dual from information_schema.system_users limit 1) as dual]

	at org.jooq_3.11.2.HSQLDB.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2380)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:811)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:364)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:393)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:380)
	at org.jooq.impl.AbstractResultQuery.fetchOne(AbstractResultQuery.java:545)
	at org.jooq.impl.AbstractResultQuery.fetchOne(AbstractResultQuery.java:481)
	at org.jooq.impl.SelectImpl.fetchOne(SelectImpl.java:2819)
	at org.jooq.impl.AbstractRoutine.executeSelect(AbstractRoutine.java:455)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:385)
	at org.jooq.impl.AbstractRoutine.execute(AbstractRoutine.java:321)
	at com.twosigma.ops.fobo.trade.data.generated.Routines.getUniqueIds(Routines.java:49)
	at com.twosigma.ops.fobo.trade.data.MyTest.canAllocateUniqueIds(MyTest.java:80)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.sql.SQLSyntaxErrorException: user lacks privilege or object not found: get_unique_ids in statement [select "dbo"."get_unique_ids"(cast(? as varchar(32672)), cast(? as int), cast(? as bigint)) from (select 1 as dual from information_schema.system_users limit 1) as dual]
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.<init>(Unknown Source)
	at org.hsqldb.jdbc.JDBCConnection.prepareStatement(Unknown Source)
	at com.twosigma.dbpool.ConnectionProxy.prepareStatement(ConnectionProxy.java:206)
	at org.jooq.impl.ProviderEnabledConnection.prepareStatement(ProviderEnabledConnection.java:109)
	at org.jooq.impl.SettingsEnabledConnection.prepareStatement(SettingsEnabledConnection.java:73)
	at org.jooq.impl.AbstractResultQuery.prepare(AbstractResultQuery.java:239)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:322)
	... 33 more
Caused by: org.hsqldb.HsqlException: user lacks privilege or object not found: get_unique_ids
	at org.hsqldb.error.Error.error(Unknown Source)
	at org.hsqldb.error.Error.error(Unknown Source)
	at org.hsqldb.ParserDQL.readColumnOrFunctionExpression(Unknown Source)
	at org.hsqldb.ParserDQL.XreadSimpleValueExpressionPrimary(Unknown Source)
	at org.hsqldb.ParserDQL.XreadAllTypesValueExpressionPrimary(Unknown Source)
	at org.hsqldb.ParserDQL.XreadAllTypesPrimary(Unknown Source)
	at org.hsqldb.ParserDQL.XreadAllTypesFactor(Unknown Source)
	at org.hsqldb.ParserDQL.XreadAllTypesTerm(Unknown Source)
	at org.hsqldb.ParserDQL.XreadAllTypesCommonValueExpression(Unknown Source)
	at org.hsqldb.ParserDQL.XreadValueExpression(Unknown Source)
	at org.hsqldb.ParserDQL.XreadSelect(Unknown Source)
	at org.hsqldb.ParserDQL.XreadQuerySpecification(Unknown Source)
	at org.hsqldb.ParserDQL.XreadSimpleTable(Unknown Source)
	at org.hsqldb.ParserDQL.XreadQueryPrimary(Unknown Source)
	at org.hsqldb.ParserDQL.XreadQueryTerm(Unknown Source)
	at org.hsqldb.ParserDQL.XreadQueryExpressionBody(Unknown Source)
	at org.hsqldb.ParserDQL.XreadQueryExpression(Unknown Source)
	at org.hsqldb.ParserDQL.compileCursorSpecification(Unknown Source)
	at org.hsqldb.ParserCommand.compilePart(Unknown Source)
	at org.hsqldb.ParserCommand.compileStatement(Unknown Source)
	at org.hsqldb.Session.compileStatement(Unknown Source)
	at org.hsqldb.StatementManager.compile(Unknown Source)
	at org.hsqldb.Session.execute(Unknown Source)
	... 40 more
```
### Steps to reproduce the problem (include example code if possible):
Inserted a stored procedure using context.execute():
```
DROP PROCEDURE "dbo"."get_unique_ids" IF EXISTS;
CREATE PROCEDURE "dbo"."get_unique_ids"(
 IN p_name    VARCHAR(128),
 IN p_count INT,
 OUT ret_id    BIGINT
)
MODIFIES SQL DATA
BEGIN ATOMIC
SET ret_id = ...;
END
```

The code for the jOOQ routine was generated from an SQL Server database:
```
IF NOT EXISTS (SELECT * FROM sysobjects AS so WHERE so.name = 'unique_ids')
BEGIN
    CREATE TABLE [dbo].[unique_ids] (...) ON [PRIMARY]
END

IF NOT EXISTS (SELECT * FROM sysobjects AS so WHERE so.name = 'get_unique_ids')
BEGIN
    EXEC('CREATE PROCEDURE dbo.get_unique_ids (
               @name   VARCHAR(128),
               @count  INT,
               @id BIGINT OUTPUT)
        AS
        BEGIN
            SET @id = ...
        END')
END
```

Call the procedure using the jOOQ routine:
```
DSLContext context = // build my context
Long out = null;
Routines.getUniqueIds(context.configuration(), "test", 4, out);
```

I am not sure what all the "dual" table queries are about, so not sure how to debug this. Please let me know! I'm able to call the stored procedure using the JDBC connection directly.

### Versions:

jOOQ: 3.11.2
Java: 1.8.0_172
OS: Linux
JDBC Driver (include name if inofficial driver): HSQLDB

**Comment by lukaseder** [comment]
Thank you very much for your report. Interesting, the most likely reason for this is that in SQL Server, all procedures have a return value like functions. This is why your procedure got generated as `AbstractRoutine<Integer>`. The feature was added to jOOQ 3.6 with #4106.

You can turn off this code generation feature using `<ignoreProcedureReturnValues/>` in your code generator configuration. Which makes me think that perhaps #4106 should be solved differently for improved vendor agnosticity...?

Also, note that in SQL Server, there is no such thing as a pure `OUT` parameter in SQL Server, as all of them are effectively `IN OUT` parameters. HSQLDB distinguishes between the two types of parameters, so you may want to reflect that in your HSQLDB emulation as well.

**Comment by lukaseder** [comment]
Looking into this now. I can think of two ways to fix this:

- Make this SQL Server procedure return value (from #4106) a special return value, i.e. marking it as "synthetic"
- Add some information to the generated routine indicating whether it is a procedure or a function

Both of these measures could help identify whether in HSQLDB (and other dialects), such a SQL-Server-generated routine's return value should really be considered or not.

**Comment by lukaseder** [comment]
The second option seems too laborious for this immediate problem. It's worth keeping in mind for future use, if we have more usage for this information: whether a routine is really a procedure or a function.

For now, makring the return value as synthetic will suffice.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.14.0. The fix requires new internal API in jOOQ and jOOQ-meta, which is why I won't backport it to older releases. The workround using `<ignoreProcedureReturnValues/>` can be applied for more consistent results.

During the fix, it showed that even if this is fixed, the HSQLDB procedure must still use an `INOUT` parameter, instead of a pure `OUT` one for compatibility reasons, as SQL Server doesn't support pure `OUT` parameters.

**Comment by lukaseder** [comment]
While fixing this, another caveat surfaced: https://github.com/jOOQ/jOOQ/issues/10269. It was currently not possible to prevent routine names from being quoted. So the case needed to match that of SQL Server explicitly
